tf.graph_util.extract_sub_graph should raise a better error message

Hi, 

I used `tf.graph_util.extract_sub_graph` to get a subgraph reaching a particular dest node. My dest_node is say `conv2d_transpose_7`. I by mistake gave it as a string in the second parameter. Python parses String too as a list and it raised an error `AssertionError: c is not in graph`. The `extract_sub_graph` should first check if the second parameter `isinstace` of a List so that it would be convenient to debug this kind of error in the future. A screenshot is attached.

Thanks, 


<img width="613" alt="screenshot at sep 14 23-53-32" src="https://user-images.githubusercontent.com/6195312/30446912-22e74a76-99a8-11e7-81b4-3479d2e49c09.png">

--- a/tensorflow/python/framework/graph_util_impl.py
+++ b/tensorflow/python/framework/graph_util_impl.py
@@ -21,6 +21,7 @@ from __future__ import division
 from __future__ import print_function
 import copy
 import re
+import six
 
 from tensorflow.core.framework import attr_value_pb2
 from tensorflow.core.framework import graph_pb2
@@ -123,7 +124,7 @@ def extract_sub_graph(graph_def, dest_nodes):
   if not isinstance(graph_def, graph_pb2.GraphDef):
     raise TypeError("graph_def must be a graph_pb2.GraphDef proto.")
 
-  if not isinstance(dest_nodes, list):
+  if isinstance(dest_nodes, six.string_types):
     raise TypeError("dest_nodes must be a list.")
 
   edges = {}  # Keyed by the dest node name.

--- a/tensorflow/python/framework/graph_util_impl.py
+++ b/tensorflow/python/framework/graph_util_impl.py
@@ -123,10 +123,8 @@ def extract_sub_graph(graph_def, dest_nodes):
   if not isinstance(graph_def, graph_pb2.GraphDef):
     raise TypeError("graph_def must be a graph_pb2.GraphDef proto.")
 
-  # Github 13047, cast single string to list
-  if isinstance(dest_nodes, str):
-    dest_nodes = [dest_nodes]
-    logging.warning("dest_nodes should be a list.")
+  if not isinstance(dest_nodes, list):
+    raise TypeError("dest_nodes must be a list.")
 
   edges = {}  # Keyed by the dest node name.
   name_to_node_map = {}  # Keyed by node name.

--- a/tensorflow/python/framework/graph_util_impl.py
+++ b/tensorflow/python/framework/graph_util_impl.py
@@ -126,10 +126,7 @@ def extract_sub_graph(graph_def, dest_nodes):
   # Github 13047, cast single string to list
   if isinstance(dest_nodes, str):
     dest_nodes = [dest_nodes]
-    logging.warn("dest_nodes should be a list")
-
-  if not isinstance(dest_nodes, list):
-    raise TypeError("dest_nodes must be a list, while get {}.".format(dest_nodes))
+    logging.warn("dest_nodes should be a list.")
 
   edges = {}  # Keyed by the dest node name.
   name_to_node_map = {}  # Keyed by node name.

--- a/tensorflow/python/framework/graph_util_impl.py
+++ b/tensorflow/python/framework/graph_util_impl.py
@@ -123,6 +123,14 @@ def extract_sub_graph(graph_def, dest_nodes):
   if not isinstance(graph_def, graph_pb2.GraphDef):
     raise TypeError("graph_def must be a graph_pb2.GraphDef proto.")
 
+  # Github 13047, cast single string to list
+  if isinstance(dest_nodes, str):
+    dest_nodes = [dest_nodes]
+    logging.warn("dest_nodes should be a list")
+
+  if not isinstance(dest_nodes, list):
+    raise TypeError("dest_nodes must be a list, while get {}.".format(dest_nodes))
+
   edges = {}  # Keyed by the dest node name.
   name_to_node_map = {}  # Keyed by node name.
 

--- a/tensorflow/python/framework/graph_util_impl.py
+++ b/tensorflow/python/framework/graph_util_impl.py
@@ -126,7 +126,7 @@ def extract_sub_graph(graph_def, dest_nodes):
   # Github 13047, cast single string to list
   if isinstance(dest_nodes, str):
     dest_nodes = [dest_nodes]
-    logging.warn("dest_nodes should be a list.")
+    logging.warning("dest_nodes should be a list.")
 
   edges = {}  # Keyed by the dest node name.
   name_to_node_map = {}  # Keyed by node name.

--- a/tensorflow/python/framework/graph_util_test.py
+++ b/tensorflow/python/framework/graph_util_test.py
@@ -189,7 +189,6 @@ class DeviceFunctionsTest(test.TestCase):
     self.assertEqual("n5", sub_graph.node[3].name)
 
   def testExtractSubGraphWithInvalidDestNodes(self):
-    # Github issue, 13047
     graph_def = graph_pb2.GraphDef()
     n1 = graph_def.node.add()
     n1.name = "n1"

--- a/tensorflow/python/framework/graph_util_test.py
+++ b/tensorflow/python/framework/graph_util_test.py
@@ -188,6 +188,13 @@ class DeviceFunctionsTest(test.TestCase):
     self.assertEqual("n3", sub_graph.node[2].name)
     self.assertEqual("n5", sub_graph.node[3].name)
 
+  def testExtractSubGraphWithSingleStringAsDestNodes(self):
+      # Github issue, 13047
+      graph_def = graph_pb2.GraphDef()
+      n1 = graph_def.node.add()
+      n1.name = "n1"
+      self.assertIsNotNone(graph_util.extract_sub_graph(graph_def, "n1"))
+
   def testConvertVariablesToConstsWithFunctions(self):
     @function.Defun(dtypes.float32)
     def plus_one(x):

--- a/tensorflow/python/framework/graph_util_test.py
+++ b/tensorflow/python/framework/graph_util_test.py
@@ -189,11 +189,11 @@ class DeviceFunctionsTest(test.TestCase):
     self.assertEqual("n5", sub_graph.node[3].name)
 
   def testExtractSubGraphWithSingleStringAsDestNodes(self):
-      # Github issue, 13047
-      graph_def = graph_pb2.GraphDef()
-      n1 = graph_def.node.add()
-      n1.name = "n1"
-      self.assertIsNotNone(graph_util.extract_sub_graph(graph_def, "n1"))
+    # Github issue, 13047
+    graph_def = graph_pb2.GraphDef()
+    n1 = graph_def.node.add()
+    n1.name = "n1"
+    self.assertIsNotNone(graph_util.extract_sub_graph(graph_def, "n1"))
 
   def testConvertVariablesToConstsWithFunctions(self):
     @function.Defun(dtypes.float32)

--- a/tensorflow/python/framework/graph_util_test.py
+++ b/tensorflow/python/framework/graph_util_test.py
@@ -188,12 +188,13 @@ class DeviceFunctionsTest(test.TestCase):
     self.assertEqual("n3", sub_graph.node[2].name)
     self.assertEqual("n5", sub_graph.node[3].name)
 
-  def testExtractSubGraphWithSingleStringAsDestNodes(self):
+  def testExtractSubGraphWithInvalidDestNodes(self):
     # Github issue, 13047
     graph_def = graph_pb2.GraphDef()
     n1 = graph_def.node.add()
     n1.name = "n1"
-    self.assertIsNotNone(graph_util.extract_sub_graph(graph_def, "n1"))
+    with self.assertRaisesRegexp(TypeError, "must be a list"):
+      graph_util.extract_sub_graph(graph_def, "n1")
 
   def testConvertVariablesToConstsWithFunctions(self):
     @function.Defun(dtypes.float32)

